client/pluginmanager/csimanager/volume.go

```diff
@@ -83,23 +87,31 @@ func (v *volumeManager) ensureStagingDir(vol *structs.CSIVolume) (bool, string,
 	return !isNotMount, stagingPath, nil
 }
 
-// stageVolume prepares a volume for use by allocations. When a plugin exposes
-// the STAGE_UNSTAGE_VOLUME capability it MUST be called once-per-volume for a
-// given usage mode before the volume can be NodePublish-ed.
-func (v *volumeManager) stageVolume(ctx context.Context, vol *structs.CSIVolume) error {
-	logger := hclog.FromContext(ctx)
-	logger.Trace("Preparing volume staging environment")
-	existingMount, stagingPath, err := v.ensureStagingDir(vol)
-	if err != nil {
-		return err
+// ensureAllocDir attempts to create a directory for use when publishing a volume
+// and then validates that the path is not already a mount point (e.g when reattaching
+// to existing allocs).
+//
+// Returns whether the directory is a pre-existing mountpoint, the publish path,
+// and any errors that occurred.
+func (v *volumeManager) ensureAllocDir(vol *structs.CSIVolume, alloc *structs.Allocation) (bool, string, error) {
```

Comment:
(Non-blocking style issue) I feel like this API would read a little more conventionally if the `string` was before the `bool`:

```go
path, preExisting, err := ensureAllocDir()
```

I almost want to return the bool as a different error and have the caller's behavior switch on the error type (only bubbling up for the non-prexisting case). But I also get that return a non-default value + a non-nil error isn't very idiomatic.